src/libraries/System.Resources.Extensions/src/System/Resources/Extensions/BinaryFormat/Deserializer/ObjectRecordDeserializer.cs

```diff
@@ -0,0 +1,103 @@
+{
+    // Used to indicate that the value is missing from the deserialized objects.
+    private protected static object s_missingValueSentinel = new();
+
+    internal SerializationRecord ObjectRecord { get; }
+
+    [AllowNull]
+    internal object Object { get; private protected set; }
+
+    private protected IDeserializer Deserializer { get; }
+
+    private protected ObjectRecordDeserializer(SerializationRecord objectRecord, IDeserializer deserializer)
+    {
+        Deserializer = deserializer;
+        ObjectRecord = objectRecord;
+    }
+
+    /// <summary>
+    ///  Continue parsing.
+    /// </summary>
+    /// <returns>The id that is necessary to complete parsing or <see cref="Id.Null"/> if complete.</returns>
+    internal abstract Id Continue();
+
+    /// <summary>
+    ///  Gets the actual object for a member value primitive or record. Returns <see cref="s_missingValueSentinel"/> if
+    ///  the object record has not been encountered yet.
+    /// </summary>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private protected (object? value, Id id) UnwrapMemberValue(object? memberValue)
+    {
+        if (memberValue is null) // PayloadReader does not return NullRecord, just null
+        {
+            return (null, Id.Null);
+        }
+        else if (memberValue is not SerializationRecord serializationRecord) // a primitive value
+        {
+            return (memberValue, Id.Null);
+        }
+        else if (serializationRecord.RecordType is RecordType.BinaryObjectString)
+        {
+            PrimitiveTypeRecord<string> stringRecord = (PrimitiveTypeRecord<string>)serializationRecord;
+            return (stringRecord.Value, stringRecord.ObjectId);
+        }
+        else if (serializationRecord.RecordType is RecordType.MemberPrimitiveTyped)
+        {
+            return (serializationRecord.GetMemberPrimitiveTypedValue(), Id.Null);
+        }
+        else
+        {
+            // ClassRecords & ArrayRecords
+            return TryGetObject(serializationRecord.ObjectId);
+        }
+
+        (object? value, Id id) TryGetObject(Id id)
+        {
+            if (!Deserializer.DeserializedObjects.TryGetValue(id, out object? value))
+            {
+                return (s_missingValueSentinel, id);
+            }
+
+            ValidateNewMemberObjectValue(value);
+            return (value, id);
+        }
+    }
+
+    /// <summary>
+    ///  Called for new non-primitive reference types.
+    /// </summary>
+    private protected virtual void ValidateNewMemberObjectValue(object value) { }
+
+    /// <summary>
+    ///  Returns <see langword="true"/> if the given record's value needs an updater applied.
+    /// </summary>
+    private protected bool DoesValueNeedUpdated(object value, Id valueRecord) =>
+        // Null Id is a primitive of some sort.
+        !valueRecord.IsNull
+            // IObjectReference is going to have it's object replaced.
```

Comment:
```suggestion
            // IObjectReference is going to have its object replaced.
```

* [it's](https://en.wiktionary.org/wiki/it%27s#English) is a contraction of "it is", "it has", or "it was", (and maybe also "there is", or "there are").
* [its](https://en.wiktionary.org/wiki/its#English) is a possessive pronoun, the English neuter equivalent of "his" or "her"